### ISSUE TITLE
Add rule-selector-property-disallowed-list

### ISSUE THREAD

**Issue by keganlance** [stylelint/stylelint#5433]
To further limit the freedom, it would be very welcome to disallow specific properties on certain selectors.

For example, not allowing the background-color property to be set on the selector "html".

Basically the same as property-disallowed-list, but mapped on selector. So you end up with a config like this:

```
'selector-property-disallowed-list': {
  'html': ['background-color']
}
```

**Comment by jeddy3** [comment]
@keganlance Thanks for the request and for providing an example.

Sounds good to me. The `*-disallowed-list` rules are how teams enforce these types of constraints.

We've prior art for selectors in the [`selector-disallowed-list` rule](https://stylelint.io/user-guide/rules/list/selector-disallowed-list/). Like that rule, this rule should apply to the whole _selector list_.

The rule should be scoped to a _rule_ (aka _rule-set_), like the [`declaration-property-value-disallowed-list` rule](https://stylelint.io/user-guide/rules/list/declaration-property-value-disallowed-list/) is scoped to a _declaration_.

- Name: `rule-selector-property-disallowed-list`
- Primary option: `object`
- Secondary options: None
- Autofixable: No
- Message: `Unexpected property "${property}" for selector "${selector}"`
- Description: "Specify a list of disallowed properties for selectors within rules."
- Section: "Limit language features" -> "Rules"

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

You can look at the two rules I mentioned above, and the [`at-rule-property-required-list`](https://github.com/stylelint/stylelint/tree/v14/lib/rules/at-rule-property-required-list) one, for a blueprint.

Branch off `v14` and not `master`.

**Comment by aNickzz** [comment]
I'd like to request an extension to this slightly, I'd like to disallow certain properties in non-nested selectors.

```scss
.foo {
  width: 35px; //error
}

.bar .foo {
  width: 35px; //ok
}
```

My motivation for this is that there are particular properties that really depend on context. For example the `width` a button depends on its layout, as does the `margin` around a `.heading-1`. Margins especially, since they directly relate to the spacing between components rather than the component itself. The layout depends on context.

That said, I'm inclined to think that this should be valid too

```scss
.foo__bar {
  width: 100%;
}
```

Which might be achievable with some regex trickery. If this new rule were to allow regex in the selector (eg. the `'html'` part) this could be achieved with something like

```json
"selector-property-disallowed-list": {
  "/^[^ ]+$/": ["/margin.*/", "width", "height"]
}
```

There are obviously problems with that particular regex if you wanted to handle attribute selectors that contain spaces, but I suspect this kind of regex is easier to implement than a nesting level test, and a check for `__` in a class name is too specific to an opinionated code style to implement in stylelint directly.

**Comment by jeddy3** [comment]
@aNickzz Thanks for chiming in with your use case. It's always interesting to hear these.

> If this new rule were to allow regex in the selector (eg. the 'html' part)

Like the existing [`selector-disallowed-list` rule](https://stylelint.io/user-guide/rules/list/selector-disallowed-list/), this new rule should accept regex for both the selector and property parts.

> I'd like to disallow certain properties in non-nested selectors... For example the width a button depends on its layout, as does the margin around a .heading-1. Margins especially, since they directly relate to the spacing between components rather than the component itself. The layout depends on context.

This is the perfect candidate for [a plugin](https://stylelint.io/developer-guide/plugins/). You'll have much more control to craft something to your needs, whereas this rule will cater to easy and common use cases. For example, in a plugin, you'll be able to use selector parser to separate out attribute, class and nesting selectors.

**Comment by OldStarchy** [comment]
I'll look into it, thanks!
